### ISSUE TITLE
CRI: Should kubelet construct container "names" or not?

### ISSUE THREAD

**Issue by yujuhong** [kubernetes/kubernetes#30463]
Forked from the discussion in #30083 for implementing against CRI. I remember there is an old issue for this, but couldn't find it anyway.

It's worth noted that changing the container naming format is likely to cause one-time backward compatibility issues during upgrades, and may need non-trivial effort to ensure nothing would break. We'll need to determine whether it's worth the efforts to overhaul the naming logic right now. However, it shouldn't be a blocker for the ongoing kuberuntime implementation. We can use this issue to discuss what naming format to use and whether we want to make the change and/or make it backward compatible.

Kubelet embeds a lot of information in the (docker) container name itself so that it can retrieve the information easily. The current format:
`k8s_<container_name>.<hash_of_container>_<pod_name>_<pod_namespace>_<pod_UID>_<random_ID>`

* kubelet detects whether a container meets the desired spec by checking the hash of the container.
* As for the random ID, it's used to ensure each name is unique since kubelet creates a new instance of the container every time a container terminates (i.e., restarting).
* The "k8s_" prefix is used to detect whether this container is managed by kubelet or not.

E.g.,
```
pod name: heapster-v1.1.0-527143062-67der
pod namespace: kube-system
container name: heapster-nanny

The docker container name:
k8s_heapster-nanny.1a2e29e8_heapster-v1.1.0-527143062-67der_kube-system_b288c37f-58d9-11e6-a5b5-42010af00002_d4cdbac9
```

With CRI, there is no need to construct a super long container name. We could either rely more on the labels, or potentially rely on the runtime to provide more information. 
However, we still need to 1) make container name meaningful and user-friendly, and 2) generate unique name 

IMO, we can write the `<hash_of_container>` as a label and remove it from the name. The pod UID, though useful in some cases (when pods get recreated), is too long and should be omitted. The new name will look like:
 `k8s_<container_name>_<pod_name>_<pod_namespace>_<random_ID>`

@kubernetes/sig-node @thockin @dchen1107

**Comment by yifan-gu** [comment]
Ref my comments: https://github.com/kubernetes/kubernetes/pull/30083#discussion_r74478291

> My point is such logic can live in the runtime shim code, e.g. the docker shim can construct the name with such way so nothing changed as today. and for rkt, it construct the annotation to include hash, namespace, etc.
When listing containers, the runtime should be able to return only containers that are managed by kubelet (e.g. docker list with the name filter, rkt list with the annotation filter), on top of that, runtimes can still apply whatever filters that kubelet passes to them.
I am generally against putting things into names, especially if kubelet forces all runtimes to do so. And as a result, we will have to maintain such naming scheme in the future to keep backwards compatibility, that's why I am saying this is an implicit contract.
Although I agree it could be somehow helpful for users to know what containers are managed by kubelet easier.


Also, let the runtime decide how to name the container is good for backward compatibility, as kubelet doesn't need to carry the code that maintains backward compatibility.

Btw, what's the length limit of the docker name, pod name, pod namespace? Do we need to take that into consideration?

**Comment by euank** [comment]
@yifan-gu last time I tested the name length limit on docker I think it was several MB that it started breaking here-and-there, but that was docker 1.6. Note that you have to write directly to the docker.sock since most shells have a length limit... I verified at least that docker 1.11 still has a quite generous limit (`docker run --name=$(ruby -e 'puts "x"*1024*10') busybox` works; 10KB).

:+1: for it making more sense for backwards compatibility hacks to live in the runtime whenever possible. Also :+1: to making sure the name is human-readable / usable as the main purpose. Arguably, the name doesn't even need to be part of the interface at all and it could be up to the runtime to define that, though I think code-deduplication/consistency is worth not pushing it down.

**Comment by vishh** [comment]
If we are talking long-term, kubelet should be having its own checkpoint and stop using runtime features as checkpoints. 
Why do containers have to be human readable? Can't users use labels to identify containers?

**Comment by yifan-gu** [comment]
@vishh I don't know how you can apply label filters to  `docker list` or `rkt list`

Although the goal is users should never ssh into the machine and `docker list`

**Comment by yujuhong** [comment]
> Why do containers have to be human readable? Can't users use labels to identify containers?

It's just for convenience, and I would also be pretty lost if I run `docker ps` and only see `happy_easley`, `grave_varahamihira,` and `serene_thompson`. This could be solved if we have a CLI tool wrapping the runtime, but we don't have that yet. I don't see downsides of having a readable names for now.


> Although the goal is users should never ssh into the machine and docker list

Not for users, but yes for sys admins and developers, until we have other tools.

**Comment by yujuhong** [comment]
> If we are talking long-term, kubelet should be having its own checkpoint and stop using runtime features as checkpoints. 

Checkpoint is good and all, but I don't think labels are bad in anyway, as we can use it for grouping and easy filtering.

**Comment by yujuhong** [comment]
Hers is what I think:
   1. kubelet will mostly rely on labels/annotations to store kubernetes specific information.
   2. User-friendly names are still good until we have better tools for sys admin/developers.
   3. Changing the naming scheme now requires more efforts to ensure backward compatibility. Do we want to do it now?

The bigger question is whether kubelet or runtime should dictate the name. I have a slight preference for kubelet to determine the name for the time being, as it has more k8s information.

**Comment by yifan-gu** [comment]
Agree 1, 2, 3.

But I am reluctant to have other runtimes adopting the same container names as today. So I am +1 on let runtime decide the name.

**Comment by yifan-gu** [comment]
My reasoning is if we already think the naming today is bad, we should not then spread the naming scheme to a bigger scope.

**Comment by vishh** [comment]
@yifan-gu 

```shell
docker run -d -l foo=bar busybox sleep 100 && ^Ccker run -d -l foo=baz busybox sleep 100
docker ps -a -f label=foo
docker ps -a -f label=foo=baz
```

> Although the goal is users should never ssh into the machine and docker ps

If this is the goal then why make container names human readable?

@yujuhong 

> It's just for convenience, and I would also be pretty lost if I run docker ps and only see happy_easley, grave_varahamihira, and serene_thompson. This could be solved if we have a CLI tool wrapping the runtime, but we don't have that yet. I don't see downsides of having a readable names for now.

As I posted above, if we standardize on labels, it is a much better UX than scrolling through 100+ containers. 

> Checkpoint is good and all, but I don't think labels are bad in anyway, as we can use it for grouping and easy filtering.

Fair enough. But it is a kubelet - runtime implementation detail.

> Changing the naming scheme now requires more efforts to ensure backward compatibility. Do we want to do it now?

The naming schema was important when docker did not have labels. Given that we are requiring runtimes to support labels, is this still necessary. Peronally, I find any kubelet constructed format not human friendly since one has to know apriori the format that kubelet uses to construct container names. Using labels is much simpler and fits in with the k8s API model as well.

**Comment by yujuhong** [comment]
> As I posted above, if we standardize on labels, it is a much better UX than scrolling through 100+ containers.

I usually does `docker ps` to see what are running on the node. The fact that I have to figure out what *pods* are running on the node first and use to filter feels clunky to me. This is just my opinion though.

> The naming schema was important when docker did not have labels. Given that we are requiring runtimes to support labels, is this still necessary. Peronally, I find any kubelet constructed format not human friendly since one has to know apriori the format that kubelet uses to construct container names. Using labels is much simpler and fits in with the k8s API model as well.

Agreed that the current naming schema is not the best, but at least I can browse and see what addon pods are running. We can also improve it as I originally suggested. As for the k8s API, imagine if you run `kubectl get pods` and the output only prints a list of pod UIDs. I don't think anyone would like the experience.


> My reasoning is if we already think the naming today is bad, we should not then spread the naming scheme to a bigger scope.

@yifan-gu, I totally agree with the sentiment and I see the reasoning. However, this only implies that kubelet/k8s doesn't care about the container names at all, since runtime has completely freedom to do that. This doesn't resolve the "user friendly" name issue. I'd need to be convinced that k8s shouldn't care about the underlying container names used by the runtime to agree to this.

**Comment by yifan-gu** [comment]
@yujuhong I think if we store things in labels, then k8s doesn't care about container names in most cases, it only requires the name to be user friendly and backward compatible, which the shim code can guarantee that.

**Comment by yujuhong** [comment]
> which the shim code can guarantee that.

If we (as kubernetes) care about the naming, we should let kubelet name the containers. Otherwise, it's completely up to the shims, and we should not have unfounded expectations :-)

**Comment by Random-Liu** [comment]
> I think if we store things in labels, then k8s doesn't care about container names in most cases, it only requires the name to be user friendly and backward compatible, which the shim code can guarantee that.

@yifan-gu  If we let the shim code handle all these. What should kubelet pass to runtime when starting a container? The container name in `api.Container`? I think that we be more misleading, we have to enforce the shim to understand whether it should report error because kubelet is trying to create the same container again or it should always create a randomized name and create the container.

**Comment by yifan-gu** [comment]
> If we (as kubernetes) care about the naming, we should let kubelet name the containers. Otherwise, it's completely up to the shims, and we should not have unfounded expectations :-)

@yujuhong I don't understand why we can't expect the runtime shim to generate user friendly container names... Why runtime shim will create crazy names for no good reasons... 

> @yifan-gu If we let the shim code handle all these. What should kubelet pass to runtime when starting a container? The container name in api.Container? I think that we be more misleading, we have to enforce the shim to understand whether it should report error because kubelet is trying to create the same container again or it should always create a randomized name and create the container.

Report what error? I think the shim will always create a container with a random ID even kubelet only gives the `api.Container.Name`.

I'd rather we parse info from labels/annotations than the name string. But if you all feel good about kubelet controls the names, I won't block.

**Comment by yujuhong** [comment]
> Report what error? I think the shim will always create a container with a random ID even kubelet only gives the api.Container.Name.

Yeah..I don't think this would be a problem.

> @yujuhong I don't understand why we can't expect the runtime shim to generate user friendly container names... Why runtime shim will create crazy names for no good reasons...

Not crazy names necessarily, but we also shouldn't have the illusion that the names will be meaningful to k8s admin/developers if we didn't explicitly define those. 
I'd like to be convinced that names do not matter at all, and if that's the case, runtime can take over naming completely (I think @vishh is in this camp). Otherwise, it's easier for kubelet to unify the naming.

**Comment by vishh** [comment]
Are we basing this discussion on docker? Is `docker ps` the precedent we
are setting for all runtimes?

**Comment by yujuhong** [comment]
> Are we basing this discussion on docker? Is `docker ps` the precedent we
> are setting for all runtimes?

No, not just docker. Every runtime we are about supports naming and listing.

**Comment by yujuhong** [comment]
To be clear, I am only against this (letting runtime handles names) due to the lack of user-friendly tools to interact with container runtime (and/or kubelet) on the node.

**Comment by feiskyer** [comment]
> To be clear, I am only against this (letting runtime handles names) due to the lack of user-friendly tools to interact with container runtime (and/or kubelet) on the node.

Agree. Besides this, letting kubelet manage container's name has other benifits:

- Easy for debugging: if there is something wrong, `k8s_<container_name>_<pod_name>_<pod_namespace>_<random_ID>` is much easier to find than `<container_name>_<random_ID>`
- Uniformed sandbox/container name accrossing runtimes, so it doesn't confuse users when runtime is changed.

**Comment by vishh** [comment]
@dchen1104 clarified this issue a bit yesterday. There are two problems
combined in this issue.

1. Kubelet needing a key to identify containers via the CRI - Kubelet can
either use a human friendly name or a UUID or whatever; it doesn't really
matter to the end user.

2. Runtimes exposing the container name to humans via their UX - I don't
think kubelet should influence the UX of runtimes. Docker runtime can
choose to re-use the CRI container name as the container alias. Other
runtimes can choose to expose whatever container name they want. They can
even choose to hide containers and only expose pods. If a runtime provides
a pod level UX, the CRI level container name is irrelevant.

Also in the entirety of time, we'd ideally provide a runtime independent
tool that users can use to identify running pods and containers without
having to bother about CRI level container names.

Thoughts?

**Comment by yujuhong** [comment]
> Also in the entirety of time, we'd ideally provide a runtime independent
> tool that users can use to identify running pods and containers without
> having to bother about CRI level container names.

I agree that this is the ideal situation, as I've mentioned a couple of times in the previous comment. While we are working towards that, let's at least ensure that there is no apparent regression in user experience today.
 
> 1. Kubelet needing a key to identify containers via the CRI - Kubelet can
> either use a human friendly name or a UUID or whatever; it doesn't really
> matter to the end user.

Hmm...yes, but most calls in CRI rely on container ID, instead of names. I think name scan truly be regarded as auxiliary information for users. 

> 2. Runtimes exposing the container name to humans via their UX - I don't
> think kubelet should influence the UX of runtimes. Docker runtime can
> choose to re-use the CRI container name as the container alias. Other
> runtimes can choose to expose whatever container name they want. They can
> even choose to hide containers and only expose pods. If a runtime provides
> a pod level UX, the CRI level container name is irrelevant.

I don't think we can influence the UX of the runtimes anyway, but "name" is a common feature that most tools offers and exposes in their UX. Without any custom tool for inspection today, maintaining a readable name seems reasonable/desirable to me.

Again, IMO, names are not essential to CRI. I view it as a bridge to provide better user experience, until we have a runtime-independent tool for users. It's nice-to-have for now. 

I am going to re-title the issue to reflect our discussion.

**Comment by dchen1107** [comment]
@vishh, @yujuhong and I had an offline discussion, and agreed upon 

* No one is happy with today's name schema, which was introduced to Kubelet due to the legacy reason, but that is a separate issue. We might or might not clean up that mess through CRI work. 
* Then there are two separate issues based on above agreement which is listed at https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239532962
  * Use container name as the key to identify a container within a pod. Kubernetes has a contact with the user at API level, at any given time, each container in a pod must have a unique name: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1108 And Kubernetes enforce the contract through validation. Now how we are going to enforce the same contact between Kubelet and Container Runtime? Especially we made this client-server mode, and changed the API from declarative to imperative. 

    I expected the runtime to enforce the same contact here. For example, a container c within pod p is still running, Kubelet send runtime to start another instance of c due to a bug.  Runtime should reject this request with a proper reason. Of course we can achieve this through different way. For example, Kubelet can generate a UUID matching each container in pod as the key. But what's the benefit by introducing another key? Of course Runtime still need to maintain another unique index / key for each container instance for a given container. 

  *  Having runtime exposing container name passing from Kubernetes has nothing to do with UX, but it will help with debuggability and readability of logging. But this is not the main reason in my opinion.

**Comment by yujuhong** [comment]
> Use container name as the key to identify a container within a pod.

@dchen1107 I know I've agreed to this, but after thinking more about it, the semantics are a little bit different from how it is in the kubernetes API. The name will only be unique among "running" containers. The runtime will reject the creation request if there is already a running container with the same name, but it'd accept the request if there are only exited containers with that name. This differs from what a unique key is used in most API, and I'm a bit unsure about the special case.

**Comment by thockin** [comment]
Jumping in late, off-the-cuff thoughts.

The actual name of the container should be the runtime's decision.  We don't want to know what characters are allowed or what length-limit exists in each runtime.  We should encourage runtimes to:
a) use names such that a human has a chance of figuring out which containers are kube
b) use names such that hand-started containers don't conflict with kube ones

We need a name to pass over CRI, but that name is (namespace, podname, containername).  Starting a container with a given value of that tuple should be idempotent, right?  If Docker needs to add a random number to the end to make a unique name, that is Docker's problem (the shim, I guess :)

**Comment by yujuhong** [comment]
> Starting a container with a given value of that tuple should be idempotent, right?

@thockin, I agree with most of your comment except this part.

If you attempt to create a pod with (foo, bar) as the (name, namespace) tuple, and (foo, bar) already exists, the apiserver will reject such a request regardless whether (foo, bar) is running or terminated. 
In kubelet's case, the container runtime (and the shim) will create a new container on such a request *if there is no running container of the same tuple*. This is somewhat strange to me because it ties container existence (the result of creation) to the status of the container (running or not). Perhaps this confusion will not be there if `CreateContainer` and `StartContainer` are combined into `RunContainer`, or if CRI is declarative instead of imperative.

I think I'm the only one who finds this semantics a bit weird, and I can see how this could simplify the implementation. If everyone else agrees, we can document this and move on.

**Comment by feiskyer** [comment]
By the way, there are no pod namespace or pod UID in today's CRI interface (ContainerConfig and PodSandboxConfig). I was thinking they are included in `name` by initial. I think they are expected in sandbox/container name for easy debugging.

**Comment by dchen1107** [comment]
@yujuhong Looks like I misuse the word key in above statement (https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239566831). But we did talk about this, and I do mean at a given time, within a pod in a namespace (Thanks @thockin to add this to tuple too in below comment), there should be one and only one container with the same name running. 

@thockin, I totally agree with your comment at https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239576511 and that is the point I tried to get to. But I have the comment on idempotent part too. 

@feiskyer Yes, we should fix the missing namespace part at the interface. Thanks!

**Comment by yujuhong** [comment]
> By the way, there are no pod namespace or pod UID in today's CRI interface (ContainerConfig and PodSandboxConfig). I was thinking they are included in name by initial. I think they are expected in sandbox/container name for easy debugging.

It can still be included in the name.

**Comment by feiskyer** [comment]
> It can still be included in the name.

A problem of this is  if kubelet concat a container name and runtimes concat another name based on it, there may an underlying hypothesis that runtimes knows how kubelet concat the container name, e.g. if `_` is already used, then runtimes should use other character for concating.

**Comment by yujuhong** [comment]
> there should be one and only one container with the same name running.

Yes, that's exactly my question in https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239581550. We'd be mixing container creation and start by enforcing this. I feel like we are asking the runtime to expose primitives via CRI, while expecting it to adhere to policy across rpc calls, and I don't like that. 

That said, I think people agree on having a simpler name in CRI and let the runtime be creatively about the actual naming. I uploaded #30560, which uses `containerName_podName_podNamespace` as the new name, and let shim handle the rest. We will add the name parsing logic in dockershim for backward compatibility later.

**Comment by yujuhong** [comment]
> A problem of this is if kubelet concat a container name and runtimes concat another name based on it, there may an underlying hypothesis that runtimes knows how kubelet concat the container name, e.g. if _ is already used, then runtimes should use other character for concating.

I think it's debatable whether runtime actually needs to know the pod UID, name, and namespace. I don't see actual value of clobbering CRI with these additional fields. What is runtime going to do with it?

**Comment by thockin** [comment]
On Fri, Aug 12, 2016 at 4:13 PM, Yu-Ju Hong <notifications@github.com> wrote:
>
> Starting a container with a given value of that tuple should be idempotent, right?
>
> @thockin, I agree with most of your comment except this part.
>
> If you attempt to create a pod with (foo, bar) as the (name, namespace) tuple, and (foo, bar) already exists, the apiserver will reject such a request regardless whether (foo, bar) is running or terminated.

This is a good point.  OK.

So, do we then expect the kubelet core to keep track of "attempts" for
each container and feed that to the runtime?  E.g. "create container
ns_pod_ctr_6"?  What prevents us from accidentally creating
ns_pod_ctr_7 while 6 is still running?

Or does it make more sense to say ""create container ns_pod_ctr" and
have the runtime return an "attempt ID"?

I guess I am looking for a reason to go one way or the other.  Easier?
 Simpler?  Is there any reason that the runtime would need to know
that two containers are concrete instances of the same abstract
container?  Log aggregation?  stats?  history?  It feels to me, so
far, that CRI is a pretty low-level interface and that most
aggregation / filtering / intelligence is embedded in kubelet.  If
that's true, probably simpler is better and we should let kubelet be
smart and runtimes be dumb?

> In kubelet's case, the container runtime (and the shim) will create a new container on such a request if there is no running container of the same tuple. This is somewhat strange to me because it ties container existence (the result of creation) to the status of the container (running or not). Perhaps this confusion will not be there if CreateContainer and StartContainer are combined into RunContainer, or if CRI is declarative instead of imperative.

I thought we really wanted create and start to be discrete steps in
the lifecycle?

**Comment by thockin** [comment]
My point about naming is that we don't really know what names are allowed,
so we should keep it as simple as possible.  Instead of
`containerName_podName_podNamespace`, perhaps we should be passing a struct
/ tuple of individual tokens and letting the runtime decide what to do with
it.

**Comment by feiskyer** [comment]
> I think it's debatable whether runtime actually needs to know the pod UID, name, and namespace. I don't see actual value of clobbering CRI with these additional fields. What is runtime going to do with it?

runtimes may generate names based on those fields and they are good for debugging. For example, if it's already known there is something wrong with a pod, users don't need to filter labels/annotations to find the sandbox/containers.

> My point about naming is that we don't really know what names are allowed, so we should keep it as simple as possible.

Agree with this, if we are not tending to generate container/sandbox names inside kubelet. We can just pass the container name with Pod name/namespace/uid together to runtime.

**Comment by yujuhong** [comment]
> I guess I am looking for a reason to go one way or the other.  Easier?
>  Simpler?  Is there any reason that the runtime would need to know
> that two containers are concrete instances of the same abstract
> container?  Log aggregation?  stats?  history?  It feels to me, so
> far, that CRI is a pretty low-level interface and that most
> aggregation / filtering / intelligence is embedded in kubelet.  If
> that's true, probably simpler is better and we should let kubelet be
> smart and runtimes be dumb?

I think that was the direction we agreed on at the beginning -- letting kubelet have more control of garbage collection, log management, etc. IMO, we should stick to this and keep runtime simpler, **or** we can choose the other direction, but not both. Let me see if I can explain this more clearly with diagrams. Assume a container in a pod in k8s API is named "foo".

(A) Lifecycle of container **foo** can involve restarts.
*Deletion is omitted for simplicity.*
```                                       
              create                start          (kill or voluntarily exit)
non-existent ---------> created ----------> running -----------> terminated 
                                               ^                     | 
                                               |_____________________|
                                                              restart
```
If we want runtime to understand the state diagram above, kubelet should only speaks "foo" (not "foo_N") to the runtime, and runtime *should* implement the restart logic itself. E.g., runtime can create a new instance, or simply restart the container, and kubelet should not care how it's implemented. The CRI should include `Create, Start/Restart, Stop, Delete` calls against "foo".

(B) Lifecycle of container **foo_N**, which is an instance of container **foo**, has monotonically progressing phases.
*Deletion is omitted for simplicity.*
```

              create                start          (kill or voluntarily exit)
non-existent ---------> created ----------> running -----------> terminated

```
If we choose to implement (B) in CRI, kubelet will explicitly create a distinct container instance **foo_N**, and the runtime should not need to be aware of what "foo" is in this context. This is how we integrate with docker today.


I see values in both (A) and (B), but with how kubelet manage resources today, (B) gives kubelet more control and keeps the required changes minimal. It also reduces the code duplication in the shims for different runtimes. However, to me, speaking "foo" in (B) is confusing and misleading.

**Comment by yujuhong** [comment]
> My point about naming is that we don't really know what names are allowed,
> so we should keep it as simple as possible.  Instead of
> `containerName_podName_podNamespace`, perhaps we should be passing a struct
> / tuple of individual tokens and letting the runtime decide what to do with
> it.

This is a valid point. We can provide tuples/tokens as "hints" and let the runtime construct the name, or we can define a character set to be used in the container name, and let the shim substitute the characters however it likes. The former is more flexible, but the latter is easier to refer and reason with in kubelet.

**Comment by yifan-gu** [comment]
> In kubelet's case, the container runtime (and the shim) will create a new container on such a request if there is no running container of the same tuple. 

+ 1.

@yujuhong Why kubelet must tell runtime to create **foo_N** to achieve (B)? I am always thinking kubelet just tells runtime to create (podnamespace, podname, (pod uid), containername), runtime creates a container and returns an ID. Why the operation is not idempotent? Why runtime should reject such operation?

**Comment by yifan-gu** [comment]
I don't believe a dumb runtime should reject what kubelet tells it to do. If kubelet tells is to create container with (podname, containername), it should just go and execute. 
If kubelet makes a mistake, kubelet should fix it by telling runtime to kill the unnecessary container.

**Comment by yujuhong** [comment]
> @yujuhong Why kubelet must tell runtime to create foo_N to achieve (B)? I am always thinking kubelet just tells runtime to create (podnamespace, podname, (pod uid), containername), then runtime creates a container and returns an ID. Why the operation is not idempotent? Why runtime should reject such operation?

I am not against using a name/namespace tuple as a *hint* for the runtime to construct container names (though simply passing a name might be easier as I mentioned in https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239903145). This is just one way to implement **foo_N** in option (B).

My https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239898173 about option (A) and (B) was meant to address (or you could say oppose) the "runtime should check whether a container is running and reject requests if necessary" suggestion. My point was that if runtime needs to know the state diagram in (A), it should also be able to handle container restarts.

**Comment by euank** [comment]
We're conflating two things here. Name, in docker's case, is used both as a sorta-human-readable-hint and as an idempotency token.

It might make this better if we split those two things into conceptually different items.

I think we do need idempotency tokens for our create requests, unless we want to solve the "exactly once delivery" problem :grinning:.

If we accept that to be the case, then I think the next logical step is to explicitly split out idempotency tokens into all the calls that need them... which is actually just `Create` since all the other calls can safely be retried.

Right now, `Name` is used for idempotency, but once we've removed that responsibility from it, it, then it doesn't need to do much. Pretty much just filter.

That being said, I do think that there is still some value in making it possible to provide user-friendly names. I personally think that providing these hints as either a small firstclass struct or as a series of labels that, by convention, have well-known keys makes sense. That also ends up being fairly simple for the kubelet.

What I'm specifically proposing is doing one of the following:

1) Metadata struct - https://github.com/kubernetes/kubernetes/compare/master...euank:metadata-idempotency
2) Well known labels - https://github.com/kubernetes/kubernetes/compare/master...euank:idempotency-tokens

The docker implementation can then construct whatever name it wishes (and it could match the name it does today; that should include enough info that it can), but other runtimes can do whatever they desire with that information, whether it just be annotate it or construct its own name or what have you.

Thoughts?

**Comment by yujuhong** [comment]
@euank I agree that there are two different issues mixed in the discussion, and I was trying to clarify both with my comments (but probably not very successfully)

1. I don' think we need idempotency tokens because the semantics we want from it is too strange. See my previous comments https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239581550 and https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239898173. I'd rather the runtime remains simple and just create a unique name based on the naming hints provided by kubelet.

2. I am ok with providing naming hints either via tokens or just a simple string (i.e., the not so well-received "name_podName_podNamespace"), and let the runtime be responsible for creating a unique name within the its own naming constraints. Although I have a slight preference of imposing a more strict naming rule across runtimes (or even just use the same name), but I get this may not be as portable.

**Comment by thockin** [comment]
I'm torn on idempotence.  One could argue that our API server SHOULD
implement it, but it doesn't.  In practice this will force kubelet to
always distinguish "EEXIST" from other errors and ignore it.  Not a huge
deal.  Do what is simpler.

WRT names, I think a tuple/struct gives the right balance of structure and
freedom.  Rutimes can join the tuple however they want, it just has to be
deterministic.  Yu-Ju, I don't want to steamroll anyone, but I do want to
un-stick anything that is blocked on this.  Please tell me if you think
this is a PROBLEM or just awkward.

**Comment by yujuhong** [comment]
> I'm torn on idempotence.  One could argue that our API server SHOULD
> implement it, but it doesn't.  In practice this will force kubelet to
> always distinguish "EEXIST" from other errors and ignore it.  Not a huge
> deal.  Do what is simpler.

I've discussed with @euank on slack yesterday. I think we can ensure that "create" is idempotent if we use tokens such as `(pod UID, container name or hash, attempt number)`. The attempt number increases on every container restart, and runtime doesn't have to be aware of the difference between "foo_N" and "foo" in this context. This can prevent kubelet from making mistakes in some corner cases. Of course, this is not as powerful as what @dchen1107 and you original suggested, which I still think is problematic as I said in https://github.com/kubernetes/kubernetes/issues/30463#issuecomment-239898173 What do you think? @thockin @dchen1107 


> WRT names, I think a tuple/struct gives the right balance of structure and
> freedom.  Rutimes can join the tuple however they want, it just has to be
> deterministic.  Yu-Ju, I don't want to steamroll anyone, but I do want to
> un-stick anything that is blocked on this.  Please tell me if you think
> this is a PROBLEM or just awkward.

As for naming, I think it's mainly for user-friendliness, so passing the tuples sounds good to me.

**Comment by thockin** [comment]
SGTM

**Comment by yujuhong** [comment]
OK. To summarize, there are two separate TODO items.
 1. Change CRI to pass naming hints (pod name, pod namespace, container name) to encourage the runtime to construct a user-friendly name; also need to change the dockershim to use that information.
 2. Change CRI to pass idempotency tuple for container creation. The tuple will be tentatively be (pod UID, container hash or name, restart count). 

Volunteers to take either or both of the items are welcome. I will come back to this by end of the week if any of them is not taken.

**Comment by feiskyer** [comment]
Will take 1 today.

**Comment by euank** [comment]
I'll take 2, though I think opaque strings make more sense (and we can discuss that choice on the change)

**Comment by yujuhong** [comment]
> I'll take 2, though I think opaque strings make more sense (and we can discuss that choice on the change)

Thanks! An opaque string makes sense. My mistake. I was thinking what to include in the string and didn't make it clear.

**Comment by feiskyer** [comment]
Refer part 1 at #30753.

**Comment by euank** [comment]
Part 2 at #30830, I think we can move our discussion to each of those issues respectively in lieu of this one.

**Comment by feiskyer** [comment]
This issue could be closed since #30753 already merged.
